internal/templateutil/state.go

```diff
@@ -193,3 +206,35 @@ func getTargetNamespaces(ctx context.Context, cl client.Client, targetNamespaces
 	}
 	return result, nil
 }
+
+func getSupportedTemplates(ctx context.Context, cl client.Client, accessRule hmc.AccessRule, templateKind hmc.TemplateKind) ([]hmc.SupportedTemplate, error) {
+	switch templateKind {
+	case hmc.ClusterTemplateKind:
+		var result []hmc.SupportedTemplate
+		for _, chainName := range accessRule.ClusterTemplateChains {
+			chain := &hmc.ClusterTemplateChain{}
+			err := cl.Get(ctx, types.NamespacedName{
+				Name: chainName,
+			}, chain)
```

Comment:
I suggest calling the `list` once, and then filtering required names rather than fetching the API `N` times

Besides, I think the `list` could be called just once before looping over the `rules`, in total that'd save `NxM-1` calls for each of the kinds, which might require creating 1/2 (depending on generics usage) extra small functions to list but I believe it worths it